pkg/api/datasources.go

```diff
@@ -309,6 +314,15 @@ func (hs *HTTPServer) UpdateDataSource(c *models.ReqContext) response.Response {
 		return resp
 	}
 
+	ds, err := getRawDataSourceById(c.Req.Context(), cmd.Id, cmd.OrgId)
+	if err != nil {
+		return response.Error(409, "Datasource has already been updated by someone else. Please reload and try again", err)
```

Comment:
Same handling as delete datasource? The 409/conflict is handled further down after trying to update the datasource - do we need to have it when getting a datasource as well??

```suggestion
		if errors.Is(err, models.ErrDataSourceNotFound) {
			return response.Error(404, "Data source not found", nil)
		}
		return response.Error(500, "Failed to update datasource", err)```

Reply:
We left this behavior in place in the event that the frontend was relying on getting a 409 somewhere. 

The thinking with the 400 vs a 404 is that this is an improper request due to the datasource being read only, rather than not found.

Happy to make that change, I'll defer to you.

Reply:
> You would also get a 409 when updating a datasource that did not exist.

True, but by returning a 404 should make things more consistent so I think we should take the opportunity to fix that. The difference in Grafana UI is that there will be a different error message.

> Happy to make that change, I'll defer to you.

Given above I think my suggestions makes more sense